python/test_infra/ttir_builder.py

```diff
@@ -241,275 +262,119 @@ def eltwise_proxy(
-            return builder.add(in0, in1)
-
-
-        test_add() # NOTE Called without arguments.
-    ```
-
-    which returns
-
-    ```
-        #any = #tt.operand_constraint<...>
-        module {
-            func.func @test_add(
-                %arg0: tensor<32x32xf32>,
-                %arg1: tensor<32x32xf32>
-            ) -> tensor<32x32xf32> {
-                %0 = tensor.empty() : tensor<32x32xf32>
-                %1 = "ttir.add"(%arg0, %arg1, %0) ...
-                return %1 : tensor<32x32xf32>
-            }
-        }
-    ```
-
-    Check out:
-    https://github.com/llvm/llvm-project/blob/main/mlir/test/python/dialects/tensor.py
-    """
-
-    def decorator(test_fn: Callable):
-        # test_fn should be called with no args.
-        def wrapper():
-            ctx = Context()
-            loc = Location.unknown(ctx)
-            # Instantiate builder which is passed as the last argument to
-            # `test_fn` so the user can use it to build ops.
-            builder = TTIRBuilder(ctx, loc)
-
-            with ctx, loc:
-                test_fn_input_types = [
-                    builder.ranked_tensor_type(input_shape)
-                    for input_shape in inputs_shapes
-                ]
-
-                # Wrap everything in a mlir module.
-                module = Module.create()
-
-                with InsertionPoint(module.body):
-                    # Wrap everything in a mlir function.
-                    @func.func(*test_fn_input_types, name=test_fn.__name__)
-                    def decorated_func(*inputs):
-                        # Randomly generate golden tensors for function inputs.
-                        for i in inputs:
-                            builder.generate_and_store_random_golden(i)
-
-                        return test_fn(*inputs, builder=builder)
-
-                if module_dump:
-                    print(module)
-
-                if golden_dump:
-                    builder.print_goldens()
-
-                return module
-
-        return wrapper
-
-    return decorator
-
-
-def compile_as_mlir_module(
-    *inputs_shapes: Tuple[Shape],
-    module_dump: bool = True,
-):
-    """
-    Decorator to define a MLIR module specified as a python function.
-
-    It will wrap decorated test function in a MLIR FuncOp wrapped in a MLIR
-    module, and tie arguments of that FuncOp to test function inputs. It will
-    also pass a `TTIRBuilder` object as the last argument of test function.
-
-    Arguments
-    ---------
-    inputs_shapes: Tuple[Shape]
-        Shapes of the respective ranked tensor inputs of the test function.
+    def logical_not(
+        self, in0: Operand, golden_func: Callable = torch.logical_not
+    ) -> OpView:
+        return self.eltwise_proxy(golden_func, ttir.LogicalNotOp, [in0])
 
-    module_dump: bool
-        Set to True if printout of generated MLIR module is wished.
+    def neg(self, in0: Operand, golden_func: Callable = torch.neg) -> OpView:
```

Comment:
Why did `golden_func` become a kwarg?  Isn't the golden function a property of the op?

Reply:
reason was to give users the ability to define a golden function for an op. I was thinking with this structure, even if we have some composite op, a user can define their own golden for the composite op.

Reply:
Do you have an example in mind where this might be useful?

My thinking is that the semantics of the op, composite or not, is defined by the test infra + the dialects. Add is add, exp is exp, I'm not sure there's much room or need for supplying your own golden.  If you agree, let's revert the APIs back.

Reply:
yeah I agree with your point. The infra would define all the Goldens for every op that gets bought up. Will fix